controllers/containerimagebuild_resources.go

```diff
@@ -0,0 +1,182 @@
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cib.Name,
+			Namespace: cib.Namespace,
+			Labels:    cib.Labels,
+		},
+	}
+	if err := controllerutil.SetControllerReference(cib, sa, r.Scheme); err != nil {
+		return err
+	}
+
+	return r.Create(ctx, sa)
+}
+
+func (r *ContainerImageBuildReconciler) checkRole(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	err := r.Get(ctx, types.NamespacedName{Name: cib.Name, Namespace: cib.Namespace}, &rbacv1.Role{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cib.Name,
+			Namespace: cib.Namespace,
+			Labels:    cib.Labels,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{forgev1alpha1.GroupVersion.Group},
+				Resources: []string{"containerimagebuilds"},
+				Verbs:     []string{"get"},
+			},
+			{
+				APIGroups: []string{forgev1alpha1.GroupVersion.Group},
+				Resources: []string{"containerimagebuilds/status"},
+				Verbs:     []string{"update"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"secrets"},
+				Verbs:     []string{"get"},
+			},
+		},
+	}
+	if err := controllerutil.SetControllerReference(cib, role, r.Scheme); err != nil {
+		return err
+	}
+
+	return r.Create(ctx, role)
+}
+
+func (r *ContainerImageBuildReconciler) checkRoleBinding(ctx context.Context, cib *forgev1alpha1.ContainerImageBuild) error {
+	err := r.Get(ctx, types.NamespacedName{Name: cib.Name, Namespace: cib.Namespace}, &rbacv1.RoleBinding{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	binding := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cib.Name,
+			Namespace: cib.Namespace,
+			Labels:    cib.Labels,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     cib.Name,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      rbacv1.ServiceAccountKind,
+				Name:      cib.Name,
+				Namespace: cib.Namespace,
+			},
+		},
+	}
+	if err := controllerutil.SetControllerReference(cib, binding, r.Scheme); err != nil {
+		return err
+	}
+
+	return r.Create(ctx, binding)
+}
+
+func (r *ContainerImageBuildReconciler) jobForBuild(cib *forgev1alpha1.ContainerImageBuild) (*batchv1.Job, error) {
+	commonMeta := metav1.ObjectMeta{
+		Name:      cib.Name,
+		Namespace: cib.Namespace,
+		Labels:    cib.Labels,
+	}
+	job := &batchv1.Job{
+		ObjectMeta: commonMeta,
+		Spec: batchv1.JobSpec{
+			BackoffLimit:            pointer.Int32Ptr(0),
+			ActiveDeadlineSeconds:   pointer.Int64Ptr(3600),
+			TTLSecondsAfterFinished: pointer.Int32Ptr(0),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: commonMeta,
+				Spec: corev1.PodSpec{
```

Comment:
I think we'd need to make some sort of allowance for root privileges since things like GKE don't work without it yet.

Reply:
See 2a90504